src/variable_info.hpp

```diff
@@ -16,98 +16,119 @@
+
+// Restrict alias to local scope.
+namespace
 {
-public:
-	invalid_variablename_exception() : std::exception() {}
-	const char* what() const NOEXCEPT
-	{
-		return "invalid_variablename_exception";
-	}
-};
+namespace impl = variable_info_implementation;
+}
 
-template<const variable_info_detail::variable_info_type vit>
+/** Information on a WML variable. */
+template<typename vit>
 class variable_info
 {
 public:
+	variable_info(const std::string& varname, typename impl::maybe_const<config, vit>::reference vars) NOEXCEPT;
 
-	typedef typename variable_info_detail::maybe_const<vit,config>::type config_var;
-	/// Doesn't throw
-	variable_info(const std::string& varname, config_var& vars);
-	~variable_info();
 	std::string get_error_message() const;
-	/// Doesn't throw
-	bool explicit_index() const;
-	/// might throw invalid_variablename_exception
+
+	bool explicit_index() const NOEXCEPT;
+
+	/** @throws                 invalid_variablename_exception */
 	bool exists_as_attribute() const;
-	/// might throw invalid_variablename_exception
+
+	/** @throws                 invalid_variablename_exception */
 	bool exists_as_container() const;
 
 	/**
-		might throw invalid_variablename_exception
-		NOTE:
-			If vit == vit_const, then the lifime of the returned const attribute_value& might end with the lifetime of this object.
-	*/
-	typename variable_info_detail::maybe_const<vit, config::attribute_value>::type &as_scalar() const;
-	/// might throw invalid_variablename_exception
-	typename variable_info_detail::maybe_const<vit, config>::type & as_container() const;
-	/// might throw invalid_variablename_exception
-	typename variable_info_detail::maybe_const<vit, config::child_itors>::type as_array() const; //range may be empty
+	 * If vit == vi_policy_const, then the lifetime of the returned const attribute_value& might end
+	 * with the lifetime of this object.
+	 *
+	 * @throws                  invalid_variablename_exception
+	 */
+	typename impl::maybe_const<config::attribute_value, vit>::reference as_scalar() const;
+
+	/** @throws                 invalid_variablename_exception */
+	typename impl::maybe_const<config, vit>::reference as_container() const;
+
+	/**
+	 * Range may be empty
+	 * @throws                  invalid_variablename_exception
+	 */
+	typename impl::maybe_const<config::child_itors, vit>::type as_array() const;
 
 protected:
 	std::string name_;
-	variable_info_detail::variable_info_state<vit> state_;
-	void throw_on_invalid() const;
+	impl::variable_info_state<vit> state_;
 	bool valid_;
+
+	void throw_on_invalid() const;
 	void calculate_value();
 };
 
-/// Extends variable_info with methods that can only be applied if vit != vit_const
-template<const variable_info_detail::variable_info_type vit>
-class non_const_variable_info : public variable_info<vit>, variable_info_detail::enable_if_non_const<vit>::type
+/**
+ * Additional functionality for a non-const variable_info.
+ * @todo: should these functions take a reference?
```

Comment:
Oh, I thought you were referring to the constructor. The other functions probably shouldn't take a reference, due to that note (which I think you removed) about the vector being emptied as a side-effect.